inc/c_util/constbuffer.h

```diff
@@ -57,6 +57,8 @@ MOCKABLE_INTERFACE(constbuffer,
     /*this creates a new constbuffer from a memory area*/
     FUNCTION(, CONSTBUFFER_HANDLE, CONSTBUFFER_Create, const unsigned char*, source, uint32_t, size),
 
+    FUNCTION(, CONSTBUFFER_HANDLE, CONSTBUFFER_CreateWithOffsetAndSize, const unsigned char*, source, uint32_t, offset, uint32_t, size),
```

Comment:
>CONSTBUFFER_CreateWithOffsetAndSize [](http://example.com/codeflow?start=35&length=35)

My first reaction was: can we just use CONSTBUFFER_CreateWithCustomFree?
It seems this new function wants to say the buffer starts ar a given offset in a memory area (let's call that start).
So we have some memory area starting at start and spanning n bytes.

And I understand we want to create a buffer [start + offset, start+offset+size]

One could use CONSTBUFFER_CreateWithCustomFree, pass start+offset as source argument, pass free as free function and start as context.

Reply:
Then I realized this actually copies the memory, so I guess this does a different job.

But I think we would want as much as possible to use the constructors that do not copy the memory. Copying memory is not really desired, in the past we had to make a serious push to eliminate copying the memory for the user payloads as they did turn out to be quite an interesting bottleneck at some point.

Reply:
If we need the memory copy I am fine with having this extra constructor for easier use - it is a bit easier to use than CONSTBUFFER_CreateWithCustomFree which is totally generic though.

Reply:
So, looks like we can actually [do pointer arithmetic](https://docs.microsoft.com/en-us/dotnet/api/system.intptr.add?view=net-5.0) in .NET with `IntPtr`. So then maybe we don't need a new constructor in C `CONSTBUFFER` because we can just use `CONSTBUFFER_Create` (as it was doing so till now) and just add the offset from the .NET side.

```c#
IntPtr address = IntPtr.Add(pinnedPayload.AddrOfPinnedObject(), (int)this.offset);
```

As for copying data, that seems to be the current design, i.e., anytime someone accesses the `NativeHandle` property, an alloc and buffer copy happens on the native side (albeit once per `ConstBuffer` instance). Are you proposing we change this design and have the C side just have a direct pointer to a pinned managed buffer or something?

Reply:
I'll do this. Given that we can do pointer arithmetic on `IntPtr` in .NET I'll abandon this PR. We can talk separately about whether a design change is in order for the managed `ConstBuffer` type.